### Q1

In [1]:
# Q1
import numpy as np

temperature = np.array([25, 28, 31, 35, 38, 27, 33, 40])
print("Readings:", temperature)

Readings: [25 28 31 35 38 27 33 40]


In [2]:
# Q1 a)
corrected = temperature + 2
print("After adding 2:", corrected)

After adding 2: [27 30 33 37 40 29 35 42]


In [3]:
# Q1 b)
fahrenheit = (9 / 5) * temperature + 32
print("Temperatures in Fahrenheit:", fahrenheit)

Temperatures in Fahrenheit: [ 77.   82.4  87.8  95.  100.4  80.6  91.4 104. ]


In [4]:
# Q1 c)
high = temperature[temperature > 32]
print("Readings greater than 32:", high)

Readings greater than 32: [35 38 33 40]


In [5]:
# Q1 d)
count = np.sum(temperature > 32)
print("Number of readings above 32:", count)

Number of readings above 32: 4


**Q1 e)** With vectorization, an operation like `temperature + 2` is applied to the whole array at once. NumPy does the looping internally in pre-compiled C code, so we do not pay Python's cost of type checking and interpreting every element one by one. The code is also shorter and has fewer chances for bugs.

Boolean indexing works the same way. `temperature > 32` gives an array of True/False values in a single step, and using it as an index picks out only the matching readings. Without it we would need a `for` loop with an `if` and `append`. For millions of readings the NumPy version is much faster and easier to read.

### Q2

In [6]:
# Q2
steps = np.array([[5000, 6200, 7100],
                  [8000, 7500, 9000],
                  [4500, 5100, 4800],
                  [9000, 8500, 9500]])
print(steps)

[[5000 6200 7100]
 [8000 7500 9000]
 [4500 5100 4800]
 [9000 8500 9500]]


In [7]:
# Q2 a)
print("Total steps:", np.sum(steps))

Total steps: 84200


In [8]:
# Q2 b)
print("Mean steps:", np.mean(steps))

Mean steps: 7016.666666666667


In [9]:
# Q2 c)
print("Maximum:", np.max(steps))
print("Minimum:", np.min(steps))

Maximum: 9500
Minimum: 4500


In [10]:
# Q2 d)
print("Total steps for each day:", np.sum(steps, axis=0))

Total steps for each day: [26500 27300 30400]


In [11]:
# Q2 e)
print("Total steps for each user:", np.sum(steps, axis=1))

Total steps for each user: [18300 24500 14400 27000]


In [12]:
# Q2 f)
pos = np.argmax(steps)
print("Flat index of maximum:", pos)

user = pos // steps.shape[1]
day = pos % steps.shape[1]
print("User", user + 1, "on day", day + 1, "has the maximum steps:", steps[user, day])

Flat index of maximum: 11
User 4 on day 3 has the maximum steps: 9500


`argmax()` on a 2-D array gives the position in the flattened array, so dividing by the number of columns gives the row (user) and the remainder gives the column (day).

### Q3

In [13]:
# Q3 a)
original = np.array([1, 2, 3, 4, 5, 6])
print(original)

[1 2 3 4 5 6]


In [14]:
# Q3 b)
subset = original[1:4]
print("subset:", subset)

subset: [2 3 4]


In [15]:
# Q3 c)
subset[0] = 999
print("original:", original)
print("subset:", subset)

original: [  1 999   3   4   5   6]
subset: [999   3   4]


The change in `subset` also shows up in `original`, because a basic slice is only a view of the same memory and not a separate copy.

In [16]:
# Q3 d)
copied = original[1:4].copy()
copied[0] = 500
print("original:", original)
print("copied:", copied)

original: [  1 999   3   4   5   6]
copied: [500   3   4]


This time `original` is unchanged, because `.copy()` makes a new array with its own memory.

In [17]:
# Q3 e)
matrix = np.arange(1, 13).reshape(3, 4)
print(matrix)

[[ 1  2  3  4]
 [ 5  6  7  8]
 [ 9 10 11 12]]


In [18]:
# Q3 f)
print("First row:", matrix[0])
print("Last row:", matrix[-1])
print("Second column:", matrix[:, 1])
print("Rows 1-2 and columns 2-3:")
print(matrix[0:2, 1:3])

First row: [1 2 3 4]
Last row: [ 9 10 11 12]
Second column: [ 2  6 10]
Rows 1-2 and columns 2-3:
[[2 3]
 [6 7]]


In [19]:
# Q3 g)
flat = matrix.flatten()
rav = matrix.ravel()
print("flatten():", flat)
print("ravel():", rav)

flatten(): [ 1  2  3  4  5  6  7  8  9 10 11 12]
ravel(): [ 1  2  3  4  5  6  7  8  9 10 11 12]


In [20]:
# Q3 h)
rav[0] = 100
print("ravel array:", rav)
print("matrix after changing ravel array:")
print(matrix)

ravel array: [100   2   3   4   5   6   7   8   9  10  11  12]
matrix after changing ravel array:
[[100   2   3   4]
 [  5   6   7   8]
 [  9  10  11  12]]


The original matrix is affected, because `ravel()` returns a view of the original array.

In [21]:
# Q3 i)
flat[1] = 200
print("flatten array:", flat)
print("matrix after changing flatten array:")
print(matrix)

flatten array: [  1 200   3   4   5   6   7   8   9  10  11  12]
matrix after changing flatten array:
[[100   2   3   4]
 [  5   6   7   8]
 [  9  10  11  12]]


The original matrix is not affected, because `flatten()` always returns a copy.

In [22]:
# Q3 j)
print("Shape:", matrix.shape)
print("ndim:", matrix.ndim)
print("size:", matrix.size)
print("dtype:", matrix.dtype)

Shape: (3, 4)
ndim: 2
size: 12
dtype: int64


### Q4

In [23]:
# Q4
X = np.array([[6, 70, 3],
              [5, 50, 6],
              [8, 80, 2],
              [4, 30, 8]])
y = np.array([40, 65, 30, 85])
print(X)
print(y)

[[ 6 70  3]
 [ 5 50  6]
 [ 8 80  2]
 [ 4 30  8]]
[40 65 30 85]


In [24]:
# Q4 a)
print("Shape of X:", X.shape)
print("Dimensions of X:", X.ndim)

Shape of X: (4, 3)
Dimensions of X: 2


In [25]:
# Q4 b)
print("X.T:")
print(X.T)
print("Shape of X.T:", X.T.shape)

X.T:
[[ 6  5  8  4]
 [70 50 80 30]
 [ 3  6  2  8]]
Shape of X.T: (3, 4)


**Q4 b)** The transpose swaps rows and columns. In X each row is one user and each column is one feature. In `X.T` each row is one feature (sleep, activity, stress) and each column is one user, so the shape changes from (4, 3) to (3, 4).

In [26]:
# Q4 c)
XtX = X.T @ X
print(XtX)
print("Shape:", XtX.shape)

[[  141  1430    96]
 [ 1430 14700   910]
 [   96   910   113]]
Shape: (3, 3)


In [27]:
# Q4 d)
print("Determinant:", np.linalg.det(XtX))
XtX_inv = np.linalg.inv(XtX)
print(XtX_inv)

Determinant: 753700.000000011
[[ 1.10521428 -0.09848746 -0.14581398]
 [-0.09848746  0.00891203  0.01190129]
 [-0.14581398  0.01190129  0.0368847 ]]


The inverse exists only for a square matrix with a non-zero determinant. `X.T @ X` is 3 × 3 and its determinant is not zero, so `np.linalg.inv()` can be used here. X itself is 4 × 3, so its inverse cannot be taken.

In [28]:
# Q4 e)
beta = np.linalg.inv(X.T @ X) @ X.T @ y
print("Coefficients:", beta)

Coefficients: [-0.15921454  0.14083853 10.06302242]


**Q4 f)** There are three coefficients, one for each feature, in the order sleep hours, activity level and stress level. Each coefficient shows how much the predicted assistance score changes when that feature goes up by 1 and the other two stay the same. A positive coefficient means the score rises with that feature, and a negative one means it falls. A larger absolute value means the feature has a bigger effect on the score. Since no column of ones was added to X, this model has no intercept term.

Here the coefficients come out as about -0.16 for sleep, 0.14 for activity and 10.06 for stress. So stress level has by far the biggest effect: each extra stress point raises the predicted assistance score by about 10. Activity has a small positive effect and sleep a very small negative one. With only 4 users this is just a small practice example, so these values should not be taken as a real finding.

In [29]:
# Q4 g)
new_user = np.array([5, 40, 7])
prediction = new_user @ beta
print("Predicted assistance score:", prediction)

Predicted assistance score: 75.27862544779134
